### DISCUSSION TITLE
Testing React component - SwapDialog which calls sendTransaction

### DISCUSSION THREAD

**Discussion by pavel-lens** [wevm/wagmi#774] [Q&A] [answered]
Hi @tmm 👋 

I would really appreciate if you can give a look at following problem with testing and point me to the right direction.

**Context**
I have `SwapTokensModal.tsx` which uses `useSendTransaction()` from `wagmi`. I am trying to write a full E2E test where user clicks on the button and swaps ETH to USDC (performing any transaction is satisfactory, eg. sending ETH between 2 wallets).

I copied your code from the following files:
https://github.com/wagmi-dev/wagmi/tree/main/packages/core/test
https://github.com/wagmi-dev/wagmi/tree/main/packages/react/test

And this is how my test looks like

```tsx
export function useAccountWithConnectAndDisconnect(config: UseAccountConfig = {}) {
  return {
    account: useAccount(config),
    connect: useConnect(),
    disconnect: useDisconnect(),
  };
}

it('swaps ETH to USDC after clicking on "Swap tokens"', async () => {
    const utils = renderHook(() => useAccountWithConnectAndDisconnect()); // renderHooks() is from your `../react/testing`
    const { result, waitFor } = utils;

    await actConnect({ utils });
    await waitFor(() => expect(result.current.account.isConnected).toBeTruthy()); // works

    console.log(result.current.account.address); // works fine, displays: 0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266 (first address from getSigners()

    const component = render(
      wrapper({
        children: <SwapTokensModal isOpen={true} assetPosition={assetFrom} assetTo={assetTo} onClose={() => {}} />,
      })
    ); // renders correctly

    // Component is fetching price quote
    await component.getByText('Fetching quote');
    expect(component).toMatchSnapshot();

    // Case: need token approval = false
    await waitFor(() => !!component.getByText('Preparing swap tokens')); // works
    await waitFor(() => !!component.getByText('Swap tokens')); // works

    const btnSwapTokens = await component.getByText('Swap tokens'); // works
    await waitFor(() => expect(component.getByText('Swap tokens').getAttribute('disabled')).not.toBe('disabled'), {
      timeout: 3000,
    }); // works

    /* ====== PROBLEM AFTER CALLING THIS ====== */ 
    fireEvent.click(btnSwapTokens); // clicking on the button calls `sendTransaction()` from `useSendTransaction()`

    await waitFor(() => !!component.getByText('Tokens swapped!'), { timeout: 3000 });
  });
```

1. Wallet is connected correctly
2. Component received a price quote from API correctly, changes state to "Swap tokens" button correctly
3. Problem => After clicking the button which triggers `sendTransaction()` method from `useSendTransaction()` hook, I get the following error

```sh
ConnectorNotFoundError: Connector not found
        at Object.sendTransaction (/Users/pavel/Projects/frontend/node_modules/.pnpm/@wagmi+core@0.4.9_5bk2dx5ulmkq37vqeouh7r5vie/node_modules/@wagmi/core/dist/wagmi-core.cjs.dev.js:1126:25)
        at Object.mutationFn (/Users/pavel/Projects/frontend/node_modules/.pnpm/wagmi@0.5.10_bdyypgucfrlob5zkpcdablpgry/node_modules/wagmi/dist/wagmi.cjs.dev.js:2009:15)
        at Object.fn (/Users/pavel/Projects/frontend/node_modules/.pnpm/react-query@4.0.0-beta.23_sfoxds7t5ydpegc3knd667wn6m/node_modules/react-query/lib/core/mutation.js:91:31)
        at run (/Users/pavel/Projects/frontend/node_modules/.pnpm/react-query@4.0.0-beta.23_sfoxds7t5ydpegc3knd667wn6m/node_modules/react-query/lib/core/retryer.js:112:31)
        at createRetryer (/Users/pavel/Projects/frontend/node_modules/.pnpm/react-query@4.0.0-beta.23_sfoxds7t5ydpegc3knd667wn6m/node_modules/react-query/lib/core/retryer.js:158:5)
        at executeMutation (/Users/pavel/Projects/frontend/node_modules/.pnpm/react-query@4.0.0-beta.23_sfoxds7t5ydpegc3knd667wn6m/node_modules/react-query/lib/core/mutation.js:85:49)
        at Mutation.execute (/Users/pavel/Projects/frontend/node_modules/.pnpm/react-query@4.0.0-beta.23_sfoxds7t5ydpegc3knd667wn6m/node_modules/react-query/lib/core/mutation.js:140:26)
```

What's not clear to me @tmm is how the wallet is successfully connected at the beginning of the test via `actConnect({ utils })`, but when the component calls `sendTransaction()` it is as if some other wallet connection was actually used - which is not configured or connected.

Any pointers / ideas how to debug this would be much appreciated! I'm looking through the code of `wagmi` repo, but I can't figure this out.

Thanks a lot!

```ts
export function setupClient(config: Config = {}) {
  return createClient<Provider, WebSocketProvider>({
    connectors: [
      new MockConnector({
        chains: allChains,
        options: {
          signer: getSigners()[0]!,
        },
      }),
    ],
    provider: ({ chainId }) => getProvider({ chainId, chains: config.chains }),
    ...config,
  });
}

export function wrapper(props: Props) {
  const client = props.client ?? setupClient({ queryClient });

  return (
    <WagmiConfig client={client} {...props} />;
  );
}

```

**Comment by tmm** [comment, marked as answer]
Looks like the issue here is that you are rendering two separate trees for testing:

1. with the account, connect, and disconnect hooks using `renderHook`
2. with the app interface you want to test (e.g. `useSendTransaction`) using `render`

wagmi's `renderHook` only renders wagmi hooks inside a shell component for testing in isolation and are not meant for testing components. The best thing to do is create a [custom `render` function](https://testing-library.com/docs/react-testing-library/setup#custom-render) for your app that includes wagmi-related set up.

I created a quick example of what this might look like [here](https://github.com/tmm/testing-wagmi) (also includes end-to-end tests with [Playwright](https://playwright.dev)). Hope this helps!

**Reply by kasparkallas** [reply]
Thanks for the testing example project!

Do I understand correctly that the end-to-end tests are also using the custom render function? This assumes that there's control over the app's build and the tests are ran locally with a local blockchain.

When coming up with a testing strategy for our app, we figured it would be beneficial to run e2e tests against testnets on deployed staging environments. We felt that's the closest we could get to meaningful realistic tests.

Taking this approach, we think the intuitive path would be to set the `window.ethereum` object to a [EIP-1193](https://eips.ethereum.org/EIPS/eip-1193) ethers.js provider that we could programmatically control to sign and reject transctions. The ethers project has an [experimental EIP-1193 bridge](https://docs.ethers.io/v5/api/experimental/#experimental-eip1193bridge) to achieve this. wagmi's `InjectedConnector` manages to pick it up from `window.ethereum`.

Using this approach, we're having issues signing transactions. We're getting `chainId address mismatch` errors which I've seen happen with real wallets (Rainbow) as well:
<img width="356" alt="image" src="https://user-images.githubusercontent.com/10894666/185642556-591ca480-d960-4165-a743-3c7482ab0a40.png">

Any idea why this might be happening?

Here are references to the testing approach in Uniswap's repo (which doesn't use `wagmi`):
1. [Creation of the injected provider](https://github.com/Uniswap/interface/blob/main/cypress/support/ethereum.ts)
2. [Setting `window.ethereum` to the injected provider in a Cypress test](https://github.com/Uniswap/interface/blob/main/cypress/support/e2e.ts#L39)

What do you think of the approach in general? Would you be interested in trying to implement something like this in the example project?

**Reply by kasparkallas** [reply]
Sorry for hijacking the discussion a bit with my comment. But as a quick update, decided to ditch using `Eip1193Bridge` and just added `MockConnector` conditionally to the list of connectors even in the deployed version of the application. The application checks for a special property on the window object for a signer for the connector to use. If it's not present then the connector is not included in the list of connectors.

This seems to work nicely for E2E tests in a deployed staging environment.
